File: zookeeper/java/src/com/yahoo/zookeeper/KeeperException.java
/*
 * Copyright 2008, Yahoo! Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yahoo.zookeeper;

@SuppressWarnings("serial")
public abstract class KeeperException extends Exception {

    /**
     * All non-specific keeper exceptions should be constructed via this factory method
     * in order to guarantee consistency in error codes and such.  If you know the error
     * code, then you should construct the special purpose exception directly.  That will
     * allow you to have the most specific possible declarations of what exceptions might
     * actually be thrown.
     * @param code  The error code.
     * @param path  The zookeeper path being operated on.
     * @return  The specialized exception, presumably to be thrown by the caller.
     */
    public static KeeperException create(int code, String path) {
        KeeperException r = create(code);
        r.path = path;
        return r;
    }

    /**
     * All non-specific keeper exceptions should be constructed via this factory method
     * in order to guarantee consistency in error codes and such.  If you know the error
     * code, then you should construct the special purpose exception directly.  That will
     * allow you to have the most specific possible declarations of what exceptions might
     * actually be thrown.
     * @param code The error code of your new exception.  This will also determine the
     * specific type of the exception that is returned.
     * @return  The specialized exception, presumably to be thrown by the caller.
     */
    public static KeeperException create(int code) {
        switch (code) {
            case Code.SystemError:
                return new SystemErrorException();
            case Code.RuntimeInconsistency:
                return new RuntimeInconsistencyException();
            case Code.DataInconsistency:
                return new DataInconsistencyException();
            case Code.ConnectionLoss:
                return new ConnectionLossException();
            case Code.MarshallingError:
                return new MarshallingErrorException();
            case Code.Unimplemented:
                return new UnimplementedException();
            case Code.OperationTimeout:
                return new OperationTimeoutException();
            case Code.BadArguments:
                return new BadArgumentsException();
            case Code.APIError:
                return new APIErrorException();
            case Code.NoNode:
                return new NoNodeException();
            case Code.NoAuth:
                return new NoAuthException();
            case Code.BadVersion:
                return new BadVersionException();
            case Code.NoChildrenForEphemerals:
                return new NoChildrenForEphemeralsException();
            case Code.NodeExists:
                return new NodeExistsException();
            case Code.InvalidACL:
                return new InvalidACLException();
            case Code.AuthFailed:
                return new AuthFailedException();
            case Code.NotEmpty:
                return new NotEmptyException();
            case Code.SessionExpired:
                return new SessionExpiredException();
            case Code.InvalidCallback:
                return new InvalidCallbackException();

            case 0:
            default:
                throw new IllegalArgumentException("Invalid exception code");
        }
    }

    public void setCode(int code) {
        this.code = code;
    }

    public interface Code {
        int Ok = 0;

        // System and server-side errors
        int SystemError = -1;

        int RuntimeInconsistency = SystemError - 1;

        int DataInconsistency = SystemError - 2;

        int ConnectionLoss = SystemError - 3;

        int MarshallingError = SystemError - 4;

        int Unimplemented = SystemError - 5;

        int OperationTimeout = SystemError - 6;

        int BadArguments = SystemError - 7;

        // API errors
        int APIError = -100; // Catch all, shouldn't be used other
        // than range start

        int NoNode = APIError - 1; // Node does not exist

        int NoAuth = APIError - 2; // Current operation not permitted

        int BadVersion = APIError - 3; // Version conflict

        int NoChildrenForEphemerals = APIError - 8;

        int NodeExists = APIError - 10;

        int NotEmpty = APIError - 11;

        int SessionExpired = APIError - 12;

        int InvalidCallback = APIError - 13;

        int InvalidACL = APIError - 14;

        int AuthFailed = APIError - 15; // client authentication failed

    }

    static String getCodeMessage(int code) {
        switch (code) {
            case 0:
                return "ok";
            case Code.SystemError:
                return "SystemError";
            case Code.RuntimeInconsistency:
                return "RuntimeInconsistency";
            case Code.DataInconsistency:
                return "DataInconsistency";
            case Code.ConnectionLoss:
                return "ConnectionLoss";
            case Code.MarshallingError:
                return "MarshallingError";
            case Code.Unimplemented:
                return "Unimplemented";
            case Code.OperationTimeout:
                return "OperationTimeout";
            case Code.BadArguments:
                return "BadArguments";
            case Code.APIError:
                return "APIError";
            case Code.NoNode:
                return "NoNode";
            case Code.NoAuth:
                return "NoAuth";
            case Code.BadVersion:
                return "BadVersion";
            case Code.NoChildrenForEphemerals:
                return "NoChildrenForEphemerals";
            case Code.NodeExists:
                return "NodeExists";
            case Code.InvalidACL:
                return "InvalidACL";
            case Code.AuthFailed:
                return "AuthFailed";
            case Code.NotEmpty:
                return "Directory not empty";
            case Code.SessionExpired:
                return "Session expired";
            case Code.InvalidCallback:
                return "Invalid callback";
            default:
                return "Unknown error " + code;
        }
    }

    private int code;

    private String path;

    public KeeperException(int code) {
        this.code = code;
    }

    KeeperException(int code, String path) {
        this.code = code;
        this.path = path;
    }

    public int getCode() {
        return code;
    }

    public String getPath() {
        return path;
    }

    @Override
    public String getMessage() {
        if (path == null) {
            return "KeeperErrorCode = " + getCodeMessage(code);
        }
        return "KeeperErrorCode = " + getCodeMessage(code) + " for " + path;
    }

    public static class APIErrorException extends KeeperException {
        public APIErrorException() {
            super(Code.APIError);
        }
    }

    public static class AuthFailedException extends KeeperException {
        public AuthFailedException() {
            super(Code.AuthFailed);
        }
    }

    public static class BadArgumentsException extends KeeperException {
        public BadArgumentsException() {
            super(Code.BadArguments);
        }
    }

    public static class BadVersionException extends KeeperException {
        public BadVersionException() {
            super(Code.BadVersion);
        }
    }

    public static class ConnectionLossException extends KeeperException {
        public ConnectionLossException() {
            super(Code.ConnectionLoss);
        }
    }

    public static class DataInconsistencyException extends KeeperException {
        public DataInconsistencyException() {
            super(Code.DataInconsistency);
        }
    }

    public static class InvalidACLException extends KeeperException {
        public InvalidACLException() {
            super(Code.InvalidACL);
        }
    }

    public static class InvalidCallbackException extends KeeperException {
        public InvalidCallbackException() {
            super(Code.InvalidCallback);
        }
    }

    public static class MarshallingErrorException extends KeeperException {
        public MarshallingErrorException() {
            super(Code.MarshallingError);
        }
    }

    public static class NoAuthException extends KeeperException {
        public NoAuthException() {
            super(Code.NoAuth);
        }
    }

    public static class NoChildrenForEphemeralsException extends KeeperException {
        public NoChildrenForEphemeralsException() {
            super(Code.NoChildrenForEphemerals);
        }
    }

    public static class NodeExistsException extends KeeperException {
        public NodeExistsException() {
            super(Code.NodeExists);
        }
    }

    public static class NoNodeException extends KeeperException {
        public NoNodeException() {
            super(Code.NoNode);
        }
    }

    public static class NotEmptyException extends KeeperException {
        public NotEmptyException() {
            super(Code.NotEmpty);
        }
    }

    public static class OperationTimeoutException extends KeeperException {
        public OperationTimeoutException() {
            super(Code.OperationTimeout);
        }
    }

    public static class RuntimeInconsistencyException extends KeeperException {
        public RuntimeInconsistencyException() {
            super(Code.RuntimeInconsistency);
        }
    }

    public static class SessionExpiredException extends KeeperException {
        public SessionExpiredException() {
            super(Code.SessionExpired);
        }
    }

    public static class SystemErrorException extends KeeperException {
        public SystemErrorException() {
            super(Code.SystemError);
        }
    }

    public static class UnimplementedException extends KeeperException {
        public UnimplementedException() {
            super(Code.Unimplemented);
        }
    }
}

File: src/csharp/ZooKeeperNetEx/zookeeper/KeeperException.cs
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

using System;
using System.Collections.Generic;
using org.apache.utils;

namespace org.apache.zookeeper
 {
    /// <summary>
    /// ZooKeeper Base Exception
    /// </summary>
    public abstract class KeeperException : Exception
﻿    {
/**
     * All multi-requests that result in an exception retain the results
     * here so that it is possible to examine the problems in the catch
     * scope.  Non-multi requests will get a null if they try to access
     * these results.
     */
         private List<OpResult> results;
﻿        /**
     * All non-specific keeper exceptions should be constructed via
     * this factory method in order to guarantee consistency in error
     * codes and such.  If you know the error code, then you should
     * construct the special purpose exception directly.  That will
     * allow you to have the most specific possible declarations of
     * what exceptions might actually be thrown.
     *
     * @param code The error code of your new exception.  This will
     * also determine the specific type of the exception that is
     * returned.
     * @return The specialized exception, presumably to be thrown by
     * the caller.
     */

         internal static KeeperException create(int code, string path = null)
﻿        {
﻿            switch (EnumUtil<Code>.DefinedCast(code))
﻿            {
﻿                //case Code.SYSTEMERROR:
﻿                //    return new SystemErrorException();
﻿                case Code.RUNTIMEINCONSISTENCY:
﻿                    return new RuntimeInconsistencyException();
﻿                case Code.DATAINCONSISTENCY:
﻿                    return new DataInconsistencyException();
﻿                case Code.CONNECTIONLOSS:
﻿                    return new ConnectionLossException();
﻿                case Code.MARSHALLINGERROR:
﻿                    return new MarshallingErrorException();
﻿                case Code.UNIMPLEMENTED:
﻿                    return new UnimplementedException();
﻿                case Code.OPERATIONTIMEOUT:
﻿                    return new OperationTimeoutException();
﻿                case Code.BADARGUMENTS:
﻿                    return new BadArgumentsException(path);
﻿                //case Code.APIERROR:
﻿                //    return new APIErrorException();
﻿                case Code.NONODE:
﻿                    return new NoNodeException(path);
﻿                case Code.NOAUTH:
﻿                    return new NoAuthException();
﻿                case Code.BADVERSION:
﻿                    return new BadVersionException(path);
﻿                case Code.NOCHILDRENFOREPHEMERALS:
﻿                    return new NoChildrenForEphemeralsException(path);
﻿                case Code.NODEEXISTS:
﻿                    return new NodeExistsException(path);
﻿                case Code.INVALIDACL:
﻿                    return new InvalidACLException(path);
﻿                case Code.AUTHFAILED:
﻿                    return new AuthFailedException();
﻿                case Code.NOTEMPTY:
﻿                    return new NotEmptyException(path);
﻿                case Code.SESSIONEXPIRED:
﻿                    return new SessionExpiredException();
﻿                case Code.INVALIDCALLBACK:
﻿                    return new InvalidCallbackException();
﻿                case Code.SESSIONMOVED:
﻿                    return new SessionMovedException();
﻿                case Code.NOTREADONLY:
﻿                    return new NotReadOnlyException();
﻿                case Code.OK:
﻿                default:
﻿                    throw new ArgumentOutOfRangeException("code", "Invalid exception code");
﻿            }
﻿        }

﻿        
﻿        

﻿        /** Codes which represent the various KeeperException
         * types. This enum replaces the deprecated earlier static final int
         * constants. The old, deprecated, values are in "camel case" while the new
         * enum values are in all CAPS.
         */

         internal enum Code
﻿        {
﻿            /** Everything is OK */
﻿            OK = 0,

            // System and server-side errors.
            // This is never thrown by the server, it shouldn't be used other than
            // to indicate a range. Specifically error codes greater than this
            // value, but lesser than {@link #APIERROR}, are system errors.
            //
            //SYSTEMERROR = -1,

            /** A runtime inconsistency was found */
            RUNTIMEINCONSISTENCY = -2,
﻿            /** A data inconsistency was found */
﻿            DATAINCONSISTENCY = -3,
﻿            /** Connection to the server has been lost */
﻿            CONNECTIONLOSS = -4,
﻿            /** Error while marshalling or unmarshalling data */
﻿            MARSHALLINGERROR = -5,
﻿            /** Operation is unimplemented */
﻿            UNIMPLEMENTED = -6,
﻿            /** Operation timeout */
﻿            OPERATIONTIMEOUT = -7,
﻿            /** Invalid arguments */
﻿            BADARGUMENTS = -8,

            // API errors.
            // This is never thrown by the server, it shouldn't be used other than
            // to indicate a range. Specifically error codes greater than this
            // value are API errors (while values less than this indicate a
            // {@link #SYSTEMERROR}).
            //
            //APIERROR = -100,

            /** Node does not exist */
            NONODE = -101,
﻿            /** Not authenticated */
﻿            NOAUTH = -102,
﻿            /** Version conflict */
﻿            BADVERSION = -103,
﻿            /** Ephemeral nodes may not have children */
﻿            NOCHILDRENFOREPHEMERALS = -108,
﻿            /** The node already exists */
﻿            NODEEXISTS = -110,
﻿            /** The node has children */
﻿            NOTEMPTY = -111,
﻿            /** The session has been expired by the server */
﻿            SESSIONEXPIRED = -112,
﻿            /** Invalid callback specified */
﻿            INVALIDCALLBACK = -113,
﻿            /** Invalid ACL specified */
﻿            INVALIDACL = -114,
﻿            /** Client authentication failed */
﻿            AUTHFAILED = -115,
﻿            /** Session moved to another server, so operation is ignored */
﻿            SESSIONMOVED = -118,

﻿            /// <summary>
﻿            /// State-changing request is passed to read-only server </summary>
﻿            NOTREADONLY = -119
﻿        }

		private readonly Code code;

﻿        private readonly string path;
﻿        private KeeperException(Code code)
﻿        {
﻿            this.code = code;
﻿        }

﻿        private KeeperException(Code code, string path)
﻿        {
﻿            this.code = code;
﻿            this.path = path;
﻿        }

﻿        /**
     * Read the error Code for this exception
     * @return the error Code for this exception
     */

         internal Code getCode() {
        return code;
    }

    /**
     * Read the path for this exception
     * @return the path associated with this error, null if none
     */

         internal string getPath() {
        return path;
    }

         internal void setMultiResults(List<OpResult> res) {
        results = res;
    }

    /**
     * If this exception was thrown by a multi-request then the (partial) results
     * and error codes can be retrieved using this getter.
     * @return A copy of the list of results from the operations in the multi-request.
     * 
     * @since 3.4.0
     *
     */

         internal List<OpResult> getResults() {
             return results != null ? new List<OpResult>(results) : null;
         }

        /// <summary>
        /// <see cref="Code.AUTHFAILED"/>
        /// </summary>
        public class AuthFailedException : KeeperException
﻿        {
﻿            internal AuthFailedException()
﻿                : base(Code.AUTHFAILED)
﻿            {
﻿            }
﻿        }

        /// <summary>
        /// <see cref="Code.BADARGUMENTS"/>
        /// </summary>
        public class BadArgumentsException : KeeperException
﻿        {
﻿            internal BadArgumentsException(string path)
﻿                : base(Code.BADARGUMENTS, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.BADVERSION"/>
        /// </summary>
        public class BadVersionException : KeeperException
﻿        {
﻿            internal BadVersionException(string path)
﻿                : base(Code.BADVERSION, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.CONNECTIONLOSS"/>
        /// </summary>
        public class ConnectionLossException : KeeperException
﻿        {
﻿            internal ConnectionLossException()
﻿                : base(Code.CONNECTIONLOSS)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.DATAINCONSISTENCY"/>
        /// </summary>
        public class DataInconsistencyException : KeeperException
﻿        {
﻿            internal DataInconsistencyException()
﻿                : base(Code.DATAINCONSISTENCY)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.INVALIDACL"/>
        /// </summary>
        public class InvalidACLException : KeeperException
﻿        {
﻿            internal InvalidACLException()
﻿                : base(Code.INVALIDACL)
﻿            {
﻿            }

﻿            internal InvalidACLException(string path)
﻿                : base(Code.INVALIDACL, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.INVALIDCALLBACK"/>
        /// </summary>
        public class InvalidCallbackException : KeeperException
﻿        {
﻿            internal InvalidCallbackException()
﻿                : base(Code.INVALIDCALLBACK)
﻿            {
﻿            }
﻿        }

        /// <summary>
        /// <see cref="Code.MARSHALLINGERROR"/>
        /// </summary>
        public class MarshallingErrorException : KeeperException
﻿        {
﻿            internal MarshallingErrorException()
﻿                : base(Code.MARSHALLINGERROR)
﻿            {
﻿            }
﻿        }

        /// <summary>
        /// <see cref="Code.NOAUTH"/>
        /// </summary>
        public class NoAuthException : KeeperException
﻿        {
﻿            internal NoAuthException()
﻿                : base(Code.NOAUTH)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.NOCHILDRENFOREPHEMERALS"/>
        /// </summary>
        public class NoChildrenForEphemeralsException : KeeperException
﻿        {
﻿            internal NoChildrenForEphemeralsException(string path)
﻿                : base(Code.NOCHILDRENFOREPHEMERALS, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.NODEEXISTS"/>
        /// </summary>

        public class NodeExistsException : KeeperException
﻿        {
﻿            internal NodeExistsException(string path)
﻿                : base(Code.NODEEXISTS, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.NONODE"/>
        /// </summary>
        public class NoNodeException : KeeperException
﻿        {
﻿            internal NoNodeException(string path)
﻿                : base(Code.NONODE, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.NOTEMPTY"/>
        /// </summary>
        public class NotEmptyException : KeeperException
﻿        {
﻿            internal NotEmptyException(string path)
﻿                : base(Code.NOTEMPTY, path)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.OPERATIONTIMEOUT"/>
        /// </summary>
        public class OperationTimeoutException : KeeperException
﻿        {
﻿            internal OperationTimeoutException()
﻿                : base(Code.OPERATIONTIMEOUT)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.RUNTIMEINCONSISTENCY"/>
        /// </summary>
        public class RuntimeInconsistencyException : KeeperException
﻿        {
﻿            internal RuntimeInconsistencyException()
﻿                : base(Code.RUNTIMEINCONSISTENCY)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.SESSIONEXPIRED"/>
        /// </summary>
        public class SessionExpiredException : KeeperException
﻿        {
﻿            internal SessionExpiredException()
﻿                : base(Code.SESSIONEXPIRED)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.SESSIONMOVED"/>
        /// </summary>
        public class SessionMovedException : KeeperException
﻿        {
﻿            internal SessionMovedException()
﻿                : base(Code.SESSIONMOVED)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.NOTREADONLY"/>
        /// </summary>
        public class NotReadOnlyException : KeeperException
﻿        {
﻿            internal NotReadOnlyException() : base(Code.NOTREADONLY)
﻿            {
﻿            }
﻿        }
        
        /// <summary>
        /// <see cref="Code.UNIMPLEMENTED"/>
        /// </summary>
        public class UnimplementedException : KeeperException
﻿        {
﻿            internal UnimplementedException()
﻿                : base(Code.UNIMPLEMENTED)
﻿            {
﻿            }
﻿        }
﻿    }
﻿}
